### PR TITLE
Use autoconf and make_install actions more

### PR DESCRIPTION
**Pull request by peterjc**
I deliberately only covered what looked like easy cases. This also fixed the indentation in a few cases - we should lint all the XML files to a standard indentation pattern.

There are lots more recipes which could be re-factored to use these more advanced actions, making them shorter. On the other hand I do quite like the alternative ``shell_command`` action with CDATA approach to embed an entire bash script ;)

**This is untested**

Do we have a mechanism for mass testing all the dependency installations scripts? (I used to depend on the TTS for doing this)

### CODE REVIEW

File: packages/package_freetype_2_4/tool_dependencies.xml [hunk 1]
```diff
@@ -5,9 +5,7 @@
                 <action type="download_by_url" sha256sum="ef9d0bcb64647d9e5125dc7534d7ca371c98310fec87677c410f397f71ffbe3f">
                     http://download.savannah.gnu.org/releases/freetype/freetype-2.4.11.tar.bz2
                 </action>
-                <action type="shell_command">./configure --prefix=$INSTALL_DIR/freetype/</action>
-                <action type="shell_command">make</action>
-                <action type="shell_command">make install</action>
+                <action type="autoconf">--prefix=$INSTALL_DIR/freetype/</action>
```
Comment (yhoogstrate) [inline packages/package_freetype_2_4/tool_dependencies.xml]: ```$INSTALL_DIR/freetype/``` << is this not an unneccesairy long directory? freetype should already be in ```$INSTALL_DIR``` (this applies also to the set_environment actions)

File: packages/package_libpng_1_6_7/tool_dependencies.xml [hunk 2]
```diff
@@ -4,8 +4,7 @@
       <install version="1.0">
           <actions>
               <action type="download_by_url">http://downloads.sourceforge.net/project/libpng/libpng16/older-releases/1.6.7/libpng-1.6.7.tar.gz</action>
-                <action type="shell_command">./configure --prefix=$INSTALL_DIR</action>
-                <action type="shell_command">make &amp;&amp; make install</action>
+                <action type="autoconf">--prefix=$INSTALL_DIR</action>
```
Comment (bgruening) [inline packages/package_libpng_1_6_7/tool_dependencies.xml]: @peterjc autoconf sets `--prefix` automatically to $INSTALL_DIR. So `<action type="autoconf"\>` is enough.

File: packages/package_freetype_2_4/tool_dependencies.xml (hunk 1 above)
Comment (bgruening) [inline packages/package_freetype_2_4/tool_dependencies.xml]: This is a very old definition. I would keep it to not break compatibility with other definitions that have hard coded `/freetype` and we should try to avoid this in new definitions.

Comment (peterjc) [inline packages/package_freetype_2_4/tool_dependencies.xml]: I noticed that too and wondered about making a separate PR to remove the trailing ``freetype/`` from the whole recipe. The compatibility argument makes sense, perhaps we should add a comment to the XML explaining this?

File: packages/package_libpng_1_6_7/tool_dependencies.xml (hunk 2 above)
Comment (yhoogstrate) [inline packages/package_libpng_1_6_7/tool_dependencies.xml]: @peterjc Nice one! I like such clean-ups and I had fun seeing how many different ways there were to achieve an autoconf install.
@bgruening Great, that makes it even cleaner. I didn't know about this since it wasn't in https://wiki.galaxyproject.org/ToolDependenciesTagSets , I guess the wiki should get a small update too.

Comment (peterjc) [inline packages/package_libpng_1_6_7/tool_dependencies.xml]: I didn't guess that from the current wording of https://wiki.galaxyproject.org/ToolDependenciesTagSets - will update the PR and wiki in a moment...

File: packages/package_mcl_12_135/tool_dependencies.xml [hunk 3]
```diff
@@ -4,10 +4,7 @@
         <install version="1.0">
             <actions>
                 <action type="download_by_url">http://micans.org/mcl/src/mcl-12-135.tar.gz</action>
-                <action type="shell_command">
-                    ./configure --prefix=$INSTALL_DIR
-                </action>
-                <action type="shell_command">make install</action>
+                <action type="autoconf">--prefix=$INSTALL_DIR</action>
```
Comment (yhoogstrate) [inline packages/package_mcl_12_135/tool_dependencies.xml]: This one should definitely be tested. The original version did not have the ```make``` command, only ```make install```.

File: packages/package_tpp_4_6_3/tool_dependencies.xml [hunk 4]
```diff
@@ -77,8 +77,8 @@
                 <action type="shell_command">sed -i.bak -e 's/qw(.*) /(&amp;) /' ../perl/exporTPP.pl</action>
                 <!--action type="shell_command">echo '' &gt; ../perl/tpp_models.pl</action-->
 
-                <action type="shell_command">export PERL5LIB=$INSTALL_DIR/lib/perl5/:$PERL5LIB &amp;&amp; export PATH=$PERL_ROOT_DIR/bin/:$PATH &amp;&amp; export PERL5LIB=$PERL_ROOT_DIR/lib/perl5:$PERL5LIB &amp;&amp; make</action>
-                <action type="shell_command">make install</action>
+                <action type="shell_command">export PERL5LIB=$INSTALL_DIR/lib/perl5/:$PERL5LIB &amp;&amp; export PATH=$PERL_ROOT_DIR/bin/:$PATH &amp;&amp; export PERL5LIB=$PERL_ROOT_DIR/lib/perl5:$PERL5LIB</action>
+                <action type="make_install"/>
```
Comment (bgruening) [inline packages/package_tpp_4_6_3/tool_dependencies.xml]: `make_install` is not the same as `make install` it is `make && make install`

File: packages/package_graphicsmagick_1_3/tool_dependencies.xml [hunk 5]
```diff
@@ -3,14 +3,10 @@
         <install version="1.0">
             <actions>
                 <action type="download_by_url">http://sourceforge.net/projects/graphicsmagick/files/graphicsmagick/1.3.18/GraphicsMagick-1.3.18.tar.gz</action>
-
-                <action type="shell_command">./configure --prefix=$INSTALL_DIR/gmagick/ --enable-shared=yes</action>
-                <action type="shell_command">make</action>
-                <action type="shell_command">make install</action>
+                <action type="autoconf">--prefix=$INSTALL_DIR/gmagick/ --enable-shared=yes</action>
```
Comment (peterjc) [inline packages/package_graphicsmagick_1_3/tool_dependencies.xml]: @bgruening will this break if ``autoconf`` automatically does ``./configure --prefix=$INSTALL_DIR/`` or is it clever enough?

File: packages/package_piranha_1_2_1/tool_dependencies.xml [hunk 6]
```diff
@@ -18,9 +18,7 @@
             <package name="bamtools" version="2.3.0_2d7685d2ae"/>
           </repository>
         </action>
-        <action type="shell_command">./configure --prefix=$INSTALL_DIR --with-bam_tools_headers=$BAMTOOLS_ROOT_PATH/include/  --with-bam_tools_library=$BAMTOOLS_ROOT_PATH/lib/</action>
-        <action type="shell_command">make all</action>
-        <action type="shell_command">make install</action>
+        <action type="autoconf">--prefix=$INSTALL_DIR --with-bam_tools_headers=$BAMTOOLS_ROOT_PATH/include/  --with-bam_tools_library=$BAMTOOLS_ROOT_PATH/lib/</action>
```
Comment (yhoogstrate) [inline packages/package_piranha_1_2_1/tool_dependencies.xml]: This one needs to be tested too. ```make all``` has now been replaced with ```make```

File: packages/package_libpng_1_6_7/tool_dependencies.xml (hunk 2 above)
Comment (bgruening) [inline packages/package_libpng_1_6_7/tool_dependencies.xml]: @peterjc make_install: https://github.com/galaxyproject/galaxy/blob/dev/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py#L804
`autoconf`: https://github.com/galaxyproject/galaxy/blob/dev/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py#L462

I learned my documentation skills from @jmchilton  ;)

File: packages/package_pixman_0_32_6/tool_dependencies.xml [hunk 7]
```diff
@@ -8,14 +8,14 @@
...
               <action type="download_by_url" sha256sum="3dfed13b8060eadabf0a4945c7045b7793cc7e3e910e748a8bb0f0dc3e794904">http://cairographics.org/releases/pixman-0.32.6.tar.gz</action>
                 <action type="set_environment_for_install">
-                  <repository name="package_libpng_1_6_7" owner="iuc" prior_installation_required="True">
+                    <repository name="package_libpng_1_6_7" owner="iuc" prior_installation_required="True">
                       <package name="libpng" version="1.6.7" />
                     </repository>
                 </action>
-                <action type="shell_command">export LDFLAGS="-Wl,-rpath,$LIBPNG_LIB_PATH" &amp;&amp; echo $LDFLAGS &amp;&amp; ./configure --prefix=$INSTALL_DIR</action>
-                <action type="make_install" />
+                <action type="shell_command">export LDFLAGS="-Wl,-rpath,$LIBPNG_LIB_PATH" &amp;&amp; echo $LDFLAGS</action>
```
Comment (yhoogstrate) [inline packages/package_pixman_0_32_6/tool_dependencies.xml]: is ``` &amp;&amp; echo $LDFLAGS``` functional or a development artefact?

File: packages/package_mcl_12_135/tool_dependencies.xml (hunk 3 above)
Comment (peterjc) [inline packages/package_mcl_12_135/tool_dependencies.xml]: Good point. With most make files this would work, but you never know...

File: packages/package_qt_5_1/tool_dependencies.xml [hunk 8]
```diff
@@ -3,8 +3,7 @@
         <install version="1.0">
             <actions>
               <action type="download_by_url">http://download.qt-project.org/official_releases/qt/5.1/5.1.1/single/qt-everywhere-opensource-src-5.1.1.tar.gz</action>
-              <action type="shell_command">./configure -prefix $INSTALL_DIR -opensource -confirm-license -qt-xcb</action>
-              <action type="make_install" />
+              <action type="autoconf">-prefix $INSTALL_DIR -opensource -confirm-license -qt-xcb</action>
```
Comment (yhoogstrate) [inline packages/package_qt_5_1/tool_dependencies.xml]: s/-prefix/--prefix/ (@bgruening has mentioned this isn't necessary at all)

File: packages/package_tpp_4_6_3/tool_dependencies.xml (hunk 4 above)
Comment (peterjc) [inline packages/package_tpp_4_6_3/tool_dependencies.xml]: Yes, but there was a ``make`` call at the end of the previous shell command (line 80) which I removed.

File: packages/package_pixman_0_32_6/tool_dependencies.xml (hunk 7 above)
Comment (peterjc) [inline packages/package_pixman_0_32_6/tool_dependencies.xml]: It would be executed, and should end up in the ``INSTALLATION.log`` file I think. Not 100% sure.

File: packages/package_piranha_1_2_1/tool_dependencies.xml (hunk 6 above)
Comment (peterjc) [inline packages/package_piranha_1_2_1/tool_dependencies.xml]: Good point, I missed that.

File: packages/package_graphicsmagick_1_3/tool_dependencies.xml (hunk 5 above)
Comment (bgruening) [inline packages/package_graphicsmagick_1_3/tool_dependencies.xml]: Not sure. it depends what ./configure will take if you pass in the same two arguments. The first or the last?

File: packages/package_pixman_0_32_6/tool_dependencies.xml (hunk 7 above)
Comment (bgruening) [inline packages/package_pixman_0_32_6/tool_dependencies.xml]: Peter is probably right, this does not functional.

File: packages/package_qt_5_1/tool_dependencies.xml (hunk 8 above)
Comment (peterjc) [inline packages/package_qt_5_1/tool_dependencies.xml]: Assuming the original was correct to use ``-prefix`` then the current code will also add ``--prefix`` which may be bad. This needs testing to see if this configure script will accept the double minus sign...

File: packages/package_tpp_4_6_3/tool_dependencies.xml (hunk 4 above)
Comment (bgruening) [inline packages/package_tpp_4_6_3/tool_dependencies.xml]: Uh tricky and so unreadable in github.
Thanks Peter.

File: packages/package_qt_5_1/tool_dependencies.xml (hunk 8 above)
Comment (bgruening) [inline packages/package_qt_5_1/tool_dependencies.xml]: I think this is right and we have it installed here in Freiburg: 
See: http://doc.qt.io/qt-4.8/qt-embedded-crosscompiling.html

File: packages/package_freetype_2_4/tool_dependencies.xml (hunk 1 above)
Comment (yhoogstrate) [inline packages/package_freetype_2_4/tool_dependencies.xml]: The compatibility argument makes sense, but this will only become problematic if tools that depend on this (hard-coded) get an update in the tool shed.

Something like given below would work as well, but is also accessible by the logical directory and is concordant with the current iuc-tools. I have double checked the package and ```$INSTALL_DIR/freetype/``` is not part of the package, so it should be safe.
```
<action type="autoconf">--prefix=$INSTALL_DIR</action>
<action type="shell_command">ln -s $INSTALL_DIR $INSTALL_DIR/freetype/</action><!-- This is for backwards compatiblity -->
```

Comment (nsoranzo) [inline packages/package_freetype_2_4/tool_dependencies.xml]: @yhoogstrate Seems a good idea, but in the first line you don't need `--prefix=$INSTALL_DIR`

File: packages/package_graphicsmagick_1_3/tool_dependencies.xml (hunk 5 above)
Comment (yhoogstrate) [inline packages/package_graphicsmagick_1_3/tool_dependencies.xml]: It will override if 'prefix=' is within the autoconf <action type="autoconf"/> element
```
if 'prefix=' in configure_opts:
    pre_cmd = './configure %s && make && make install' % configure_opts
else:
    pre_cmd = './configure --prefix=$INSTALL_DIR %s && make && make install' % configure_opts
```
https://github.com/galaxyproject/galaxy/blob/dev/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py#L459

Comment (bgruening) [inline packages/package_graphicsmagick_1_3/tool_dependencies.xml]: Oh, someone was really smart at this time.

File: packages/package_qt_5_1/tool_dependencies.xml (hunk 8 above)
Comment (peterjc) [inline packages/package_qt_5_1/tool_dependencies.xml]: Currently https://github.com/galaxyproject/galaxy/blob/dev/lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py#L459 looks for ``prefix=`` in the optional arguments, so won't spot ``-prefix $INSTALL_DIR`` and will add ``--prefix=$INSTALL_DIR`` as well.

I think I need to revert this change and add a comment about why we're not using the ``autoconf`` action.

Comment (yhoogstrate) [inline packages/package_qt_5_1/tool_dependencies.xml]: It would make sense to allow something like this in the galaxy/tool shed code:
```<action type="autoconf" atuoprefix="False">-prefix $INSTALL_DIR</action>```

File: packages/package_freetype_2_4/tool_dependencies.xml (hunk 1 above)
Comment (peterjc) [inline packages/package_freetype_2_4/tool_dependencies.xml]: The same would apply to ``packages/package_graphicsmagick_1_3/tool_dependencies.xml`` which uses ``$INSTALL_DIR/gmagick/``

File: packages/package_qt_5_1/tool_dependencies.xml (hunk 8 above)
Comment (peterjc) [inline packages/package_qt_5_1/tool_dependencies.xml]: Good idea, filed as https://github.com/galaxyproject/galaxy/issues/922

File: packages/package_mcl_12_135/tool_dependencies.xml (hunk 3 above)
Comment (peterjc) [inline packages/package_mcl_12_135/tool_dependencies.xml]: This is important here,

```
[piranha-1.2.1] $ make
make: *** No targets specified and no makefile found.  Stop.
```

We do need the make target.

File: packages/package_piranha_1_2_1/tool_dependencies.xml (hunk 6 above)
Comment (peterjc) [inline packages/package_piranha_1_2_1/tool_dependencies.xml]: Fixed with https://github.com/peterjc/tools-iuc/commit/a14b9c7a8b436b8e9cd5360270560586ea90653a

The prianha-1.2.1 makefile did not define a default target.

File: packages/package_qt_5_1/tool_dependencies.xml (hunk 8 above)
Comment (peterjc) [inline packages/package_qt_5_1/tool_dependencies.xml]: Reverted in https://github.com/peterjc/tools-iuc/commit/8d8576cc87919f6b2dc5bf5a53749b49c057213e

**Comment by peterjc** [comment]
It looks like this needs a merge to apply cleanly, closing in favour of https://github.com/galaxyproject/tools-iuc/pull/334 where I did some clean up as well. Thanks all for the comments so far.

**Comment by nsoranzo** [comment]
@peterjc I don't mind a new PR, but you can rebase and force push the branch instead.

**Comment by peterjc** [comment]
If I rebased and force pushed to the same branch name, it would have updated this pull request (which would be nice), but also ophans all the useful comments on the old commits (doesn't it?)

**Comment by bgruening** [comment]
Force push doesn't remove comments as far as I know.

**Comment by peterjc** [comment]
OK, let's try a force push (and I'll close #334)...

File: packages/package_gnu_grep_2_14/tool_dependencies.xml [hunk 9]
```diff
@@ -6,18 +6,16 @@
...
+                <action type="download_by_url">https://github.com/bgruening/download_store/raw/master/GNU/grep-2.14.tar.bz2</action>
+                <action type="set_environment_for_install">
+                    <repository name="package_pcre_8_34" owner="iuc">
+                        <package name="pcre" version="8.34" />
+                    </repository>
+                </action>
+                <action type="autoconf">--enable-perl-regexp</action>
+                <action type="set_environment">
+                    <environment_variable name="PATH" action="prepend_to">$INSTALL_DIR/bin</environment_variable>
+                 </action>
```
Comment (nsoranzo) [inline packages/package_gnu_grep_2_14/tool_dependencies.xml]: Last line has wrong indentation.

Comment (peterjc) [inline packages/package_gnu_grep_2_14/tool_dependencies.xml]: Noted.

File: packages/package_tpp_4_6_3/tool_dependencies.xml (hunk 4 above)
Comment (peterjc) [inline packages/package_tpp_4_6_3/tool_dependencies.xml]: In the discussion https://github.com/galaxyproject/tools-iuc/pull/33 @bgruening pointed out the ``shell_command`` state would be lost, so these environment variables would be lost when ``make`` gets run within the ``make_install`` action.

So this should be reverted.

File: packages/package_pixman_0_32_6/tool_dependencies.xml (hunk 7 above)
Comment (peterjc) [inline packages/package_pixman_0_32_6/tool_dependencies.xml]: In the discussion https://github.com/galaxyproject/tools-iuc/pull/33 @bgruening pointed out the ``shell_command`` state would be lost, so these environment variables would be lost when ``make`` gets run within the ``make_install`` action.

So this should be reverted.

**Comment by bgruening** [comment]
@peterjc can you please rebase. I guess we are ready to merge this one, right?

**Comment by peterjc** [comment]
Rebased with the overly aggressive actions removed where they needed the environment variables being set via a ``shell_command`` action.

**Comment by nsoranzo** [comment]
:+1: 

@bgruening: can we remove the `wip` tag?

**Comment by peterjc** [comment]
I think so yes (removing the work-in-progress tag), after the discussions and useful feedback I think this is ready to be merged.

**Comment by bgruening** [comment]
@nsoranzo yes, I explicitly ask for a rebase because I think this is ready to be merged.
Thanks Peter!
